be/src/exprs/fpe.cpp

```diff
@@ -0,0 +1,194 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <iostream>
+#include <string>
+#include <sstream>
+#include <cmath>
+#include <vector>
+#include <ubiq/fpe/ff1.h>
+#include <iomanip>
+
+#include "exprs/fpe.h"
+#include "util/defer_op.h"
+
+namespace starrocks {
+
+const std::string_view FPE::DEFAULT_KEY = "abcdefghijk12345abcdefghijk12345";
+
+std::string_view FPE::trim_leading_zeros(const std::string_view& str) {
+    size_t start = str.find_first_not_of('0');
+    if (start == std::string::npos) {
+        return "0";
+    }
+    return str.substr(start);
+}
+
+std::string_view FPE::trim_trailing_zeros(const std::string_view& str) {
+    size_t end = str.find_last_not_of('0');
+    if (end == std::string::npos) {
+        return str;
+    }
+    return str.substr(0, end + 1);
+}
+
+Status FPE::encrypt(const std::string_view& num_str, const std::string_view& key, std::string& value, int radix = 10) {
+    int num_str_length = num_str.length();
+    std::string fixed_num_str(num_str);
+    if (num_str_length < MIN_LENGTH) {
+        std::string padding(MIN_LENGTH - num_str_length, '0');
+        fixed_num_str = padding + fixed_num_str;
+    }
+
+    int fpe_key_length = key.length();
+
+    std::vector<uint8_t> fpe_key(fpe_key_length);
+    for (size_t i = 0; i < fpe_key_length; ++i) {
+        fpe_key[i] = static_cast<uint8_t>(key[i]);
+    }
+
+    struct ff1_ctx* ctx = nullptr;
+    DeferOp op([&]{ if (ctx != nullptr) ff1_ctx_destroy(ctx);});
+
+    std::string tmp;
+    tmp.resize(num_str_length);
+    char* out = tmp.data();
+
+    int res = ff1_ctx_create(&ctx, fpe_key.data(),
+                             fpe_key_length, TWEAK, sizeof(TWEAK),
+                             0, SIZE_MAX,
+                             radix);
+    if (res != 0) {
+        return Status::RuntimeError("ff1_ctx_create failed");
+    }
+    res = ff1_encrypt(ctx, out, fixed_num_str.c_str(), NULL, 0);
+    if (res != 0) {
+        return Status::RuntimeError("ff1_encrypt failed");
+    }
+
+    std::string result(out);
+    value = result;
+
+    return Status::OK();
+}
+
+Status FPE::encrypt_num(const std::string_view& num_str,const std::string_view& key, std::string& value) {
+    std::string num_flag;
+    if (num_str[0] == '-') {
+        num_flag = "-";
+    }
+
+    size_t dot_pos = num_str.find('.');
+    std::string_view int_part = num_str.substr(0, dot_pos);
+    Status status;
+    std::string encrypted_int_part;
+    status = encrypt(int_part, key, encrypted_int_part, DEFAULT_RADIX);
+    if (!status.ok()) {
+        return Status::RuntimeError("encrypt_num int_part failed");
+    }
+
+    std::string encrypted_dec_part;
+    if (dot_pos != std::string::npos) {
+        std::string_view dec_part = num_str.substr(dot_pos + 1);
+        std::string dec_part_str(dec_part);
+        dec_part_str = "0." + dec_part_str;
```

Comment:
No need to prepend '0.' to dec_part_str.  .xxxxx (e.g. .12456) is legal float-pointing number represenation. and this code seems poor performance, so, the code should be written as follows;
```cpp
if (dot_pos != std::string::npos) {
        double dec_part_num = std::stod(num_str, dot_pos);
        dec_part_num = dec_part_num * EXPANDED;
        auto dec_int_part = static_cast<long long>(dec_part_num);
        std::string encrypted_dec_part;
        //TODO: here must provide a reasonable reserve size 
        encrypted_dec_part.resize(100);
        encrypted_dec_part[0]='.';
        // change encrypt signature to make it can accept a char* buffer and size_t* len,  
        size_t len = -1;      
        RETURN_IF_ERROR(encrypt(std::to_string(dec_int_part), key,encrypted_dec_part.data()+1, &len, DEFAULT_RADIX));
        encrypted_dec_part.resize(len);
}        
```

Reply:
use string_parser.hpp: string_to_float instead of stod,  since string_to_float can take char* as parameter to avoid string copying introduced by substr.
```cpp
    template <typename T>
    static inline T string_to_float(const char* s, int len, ParseResult* result) {
        T ans = string_to_float_internal<T>(s, len, result);
        return ans;
    }
```